src/cpp/shared/game/game.cpp

```diff
@@ -272,6 +273,75 @@ void philAI::RedistributeTroops(armyGroup *army1, armyGroup *army2, int a1, int
 	}
 }
 
+void game::InitRandomArtifacts() {
+  memset(artifactGeneratedRandomly, 0, sizeof(artifactGeneratedRandomly));
+  for (int i = 0; i < MAP_WIDTH; ++i) {
+    for (int j = 0; j < MAP_HEIGHT; ++j) {
+      const auto &tile = map.tiles[j * MAP_WIDTH + i];
+      if (tile.objType == (TILE_HAS_EVENT | LOCATION_ARTIFACT)) {
+        // Don't know why we need to divide by 2, but that's what the decompiled code says.
+        const int artifactId = static_cast<unsigned char>(tile.objectIndex) / 2;
+        artifactGeneratedRandomly[artifactId] = 1;
+      }
+    }
+  }
+}
+
+int game::GetRandomArtifactId(int allowedLevels, int allowNegatives) {
+  int winConditionArtifact = -1;
+  if (mapHeader.winConditionType == WIN_FIND_ARTIFACT) {
+    winConditionArtifact = mapHeader.winConditionArgumentOrLocX - 1;
+  }
+
+  // Cursed artifacts are allowed 70% of the time, every time.
+  const bool allowCursed = (allowNegatives && Random(0, 100) >= 30);
+
+  std::vector<int> choices;
+  const int numArtifacts = (xIsExpansionMap ? NUM_SUPPORTED_ARTIFACTS : MAX_BASE_ARTIFACT + 1);
+
+  for (int i = 0; i < numArtifacts; ++i) {
```

Comment:
What about ARTIFACT_NONE?

Reply:
The four NONEs have level "unused" (32), which won't match any of the requested levels in the game.  The callers of the this function only request some combination of treasure, minor, and/or major.

Reply:
This is a Level 3 bug ( http://www.pathsensitive.com/2018/01/the-three-levels-of-software-why-code.html ). Compare: https://chromium-review.googlesource.com/c/catapult/+/985025/5/telemetry/telemetry/internal/results/story_run.py#15

We need this to be a function which gets a random valid artifact ID, not a function which gets a random valid artifact ID so long as you happen to only request certain artifact ranges.